autoload/maktaba/extension.vim

```diff
@@ -0,0 +1,159 @@
+"
+" Plugins should use |Plugin.GetExtensionRegistry| to gain access to their own
+" extension registry, and @function(maktaba#extension#GetRegistry) to gain
+" access to an extension registry for another plugin.
+
+
+" Extension registry objects, keyed by plugin name.
+if !exists('s:registries')
+  let s:registries = {}
+endif
+
+
+""
+" @private
+" Returns an extension registry for the given plugin name, creating it if
+" needed.
+" @throws WrongType if {plugin} is not a string.
+function! maktaba#extension#GetInternalRegistry(plugin) abort
+  call maktaba#ensure#IsString(a:plugin)
+
+  if has_key(s:registries, a:plugin)
+    return s:registries[a:plugin]
+  endif
+
+  " TODO: should this contain the plugin's name?
+  " TODO: is there a better way to construct a do-nothing _validator?
+  let l:registry = {
+      \ 'AddExtension': function('maktaba#extension#AddExtension'),
+      \ 'GetExtensions': function('maktaba#extension#GetExtensions'),
+      \ 'SetValidator': function('maktaba#extension#SetValidator'),
+      \ '_internal_extensions': [],
+      \ '_external_extensions': [],
+      \ '_validator': maktaba#function#FromExpr('0'),
+      \ '_external': 0,
+      \ }
+  let s:registries[a:plugin] = l:registry
+  return l:registry
+endfunction
+
+
+" Throws NotImplemented.  Used by the inaccessible functions on the external
+" interface.
+function! s:NotAccessible() abort
+  throw maktaba#error#NotImplemented('Not accessible by external plugins.')
+endfunction
+
+
+""
+" Returns an extension registry for the given plugin name.
+" @throws WrongType if {plugin} is not a string.
+function! maktaba#extension#GetRegistry(plugin) abort
+  let l:registry = maktaba#extension#GetInternalRegistry(a:plugin)
+
+  " Return an external version of the extension registry with a more limited
+  " interface.
+  let l:registry = copy(l:registry)  " TODO: probably not worth caching this, right?
+  let l:registry._external = 1
+
+  return l:registry
+endfunction
+
+
+""
+" @dict ExtensionRegistry
+" Adds the given {extension} to this extension registry.
+"
+" The extension must be a dict.  If a validator is registered for this
+" extension registry, this function will call the validator.  Failures will
+" result in the validation error being shouted to the user (and the extension
+" will not be added).
+"
+" @throws WrongType if {extension} is not a dict.
+function! maktaba#extension#AddExtension(extension) dict abort
+  call maktaba#ensure#IsDict(a:extension)
+
+  try
+    call maktaba#function#Call(self._validator, [a:extension])
+    if self._external
+      call add(self._external_extensions, a:extension)
+    else
+      call add(self._internal_extensions, a:extension)
+    endif
+  catch
+    call maktaba#error#Shout(v:exception)
+  endtry
+endfunction
+
+
+""
+" @dict ExtensionRegistry
+" Returns the extensions that have been added to this extension registry.
+"
+" Extensions added by the plugin to its own registry are always returned
```

Comment:
That's reversed from how codefmtlib currently works. Makes sense if you want to use last-wins instead of first-wins registration, but in that case we should explicitly prescribe callers will normally want the last matching item, not the first.